internal/auth/authn/oauth2_auth.go

```diff
@@ -318,6 +332,214 @@ func (o *OAuth2Auth) callUserinfoEndpoint(ctx context.Context, token string) (ma
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/vnd.github+json")
+
+	resp, err := o.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to call GitHub introspection endpoint: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// GitHub returns 200 for valid tokens, 404 for invalid ones
+	if resp.StatusCode == http.StatusNotFound {
+		return fmt.Errorf("token is invalid or not owned by this application")
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("GitHub introspection endpoint returned status %d", resp.StatusCode)
+	}
+
+	o.log.Debugf("GitHub introspection succeeded")
+	return nil
+}
+
+// introspectJWT validates a token as a JWT using JWKS
+func (o *OAuth2Auth) introspectJWT(ctx context.Context, token string, spec api.JwtIntrospectionSpec) error {
+	o.log.Debugf("Starting JWT introspection with JWKS URL: %s", spec.JwksUrl)
+
+	// Parse JWT without validation first to fast-fail on non-JWT tokens
+	_, err := jwt.Parse([]byte(token), jwt.WithValidate(false), jwt.WithVerify(false))
+	if err != nil {
+		o.log.Debugf("Token is not JWT format: %v", err)
+		return fmt.Errorf("failed to parse JWT token: %w", err)
+	}
+
+	// Ensure JWKS cache is initialized (lazy initialization)
+	if err := o.ensureJWKSCache(ctx, spec); err != nil {
+		o.log.Errorf("JWKS cache initialization failed: %v", err)
+		return fmt.Errorf("JWKS cache initialization failed: %w", err)
+	}
+
+	// Get JWKS from cache
+	jwkSet, err := o.jwksCache.Get(ctx, spec.JwksUrl)
+	if err != nil {
+		o.log.Errorf("Failed to get JWKS from cache: %v", err)
+		return fmt.Errorf("failed to get JWK set from cache: %w", err)
+	}
+
+	// Parse and validate token with signature verification
+	// The jwksCache automatically refreshes based on WithMinRefreshInterval (15 minutes)
+	parsedToken, err := jwt.Parse([]byte(token), jwt.WithKeySet(jwkSet), jwt.WithValidate(true))
+	if err != nil {
+		o.log.Debugf("JWT validation failed: %v", err)
+		return fmt.Errorf("failed to validate JWT token: %w", err)
+	}
+
+	// Validate issuer if specified, otherwise use OAuth2ProviderSpec issuer
+	expectedIssuer := ""
+	if spec.Issuer != nil && *spec.Issuer != "" {
+		expectedIssuer = *spec.Issuer
+	} else if o.spec.Issuer != nil && *o.spec.Issuer != "" {
+		expectedIssuer = *o.spec.Issuer
+	}
+
+	if expectedIssuer != "" && parsedToken.Issuer() != expectedIssuer {
+		return fmt.Errorf("token issuer '%s' does not match expected issuer '%s'", parsedToken.Issuer(), expectedIssuer)
+	}
+
+	// Validate audience if specified, otherwise use OAuth2ProviderSpec clientId
+	var expectedAudiences []string
+	if spec.Audience != nil && len(*spec.Audience) > 0 {
+		expectedAudiences = *spec.Audience
+	} else if o.spec.ClientId != "" {
+		expectedAudiences = []string{o.spec.ClientId}
+	}
+
+	if len(expectedAudiences) > 0 {
+		audienceValid := false
+		tokenAudiences := parsedToken.Audience()
+		for _, expectedAud := range expectedAudiences {
```

Comment:
for _, expectedAud := range expectedAudiences {
    if slices.Contains(tokenAudiences, expectedAud) {
        audienceValid = true
        break
    }
}

Reply:
fixed